README.md
# Lavacharts
[![Total Downloads](https://img.shields.io/packagist/dt/khill/lavacharts.svg?style=plastic)](https://packagist.org/packages/khill/lavacharts)
[![License](https://img.shields.io/packagist/l/khill/lavacharts.svg?style=plastic)](http://opensource.org/licenses/MIT)
[![Minimum PHP Version](https://img.shields.io/badge/php-%3E%3D%205.4-8892BF.svg?style=plastic)](https://php.net/)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/kevinkhill/lavacharts?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
[![PayPayl](https://img.shields.io/badge/paypal-donate-yellow.svg?style=plastic)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=FLP6MYY3PYSFQ)

Lavacharts is a graphing / chart library for PHP5.4+ that wraps the Google Chart API

Stable:
[![Current Release](https://img.shields.io/github/release/kevinkhill/lavacharts.svg?style=plastic)](https://github.com/kevinkhill/lavacharts/releases)
[![Build Status](https://img.shields.io/travis/kevinkhill/lavacharts/master.svg?style=plastic)](https://travis-ci.org/kevinkhill/lavacharts)
[![Coverage Status](https://img.shields.io/coveralls/kevinkhill/lavacharts/master.svg?style=plastic)](https://coveralls.io/r/kevinkhill/lavacharts?branch=master)

Dev:
[![Current Release](https://img.shields.io/badge/release-dev--3.0-brightgreen.svg?style=plastic)](https://github.com/kevinkhill/lavacharts/tree/3.0)
[![Build Status](https://img.shields.io/travis/kevinkhill/lavacharts/3.0.svg?style=plastic)](https://travis-ci.org/kevinkhill/lavacharts)
[![Coverage Status](https://img.shields.io/coveralls/kevinkhill/lavacharts/3.0.svg?style=plastic)](https://coveralls.io/r/kevinkhill/lavacharts?branch=3.0)

## Version 3.0 is still a work in progress, but mostly stable.
Check here for notes on how to  [upgrade from 2.5.x to 3.0.x](https://github.com/kevinkhill/lavacharts/wiki/Upgrading-from-2.5-to-3.0)

## Package Features
- Blade template extensions for laravel
- Lava.js module for interacting with charts client-side
  - AJAX data reloading
  - Fetching charts
  - Events integration
- DataTable addColumn aliases
- DataTable column formatters
- [Carbon](https://github.com/briannesbitt/Carbon) support for date columns
- Supports string, number, date, and timeofday columns
- Now supporting 12 Charts!
  - Area, Bar, Calendar, Column, Combo, Donut, Gauge, Geo, Line, Pie, Scatter, Table
- [DataTablePlus](https://github.com/kevinkhill/datatableplus) package can be added to parse CSV files or Eloquent collections into DataTables.

## For complete documentation, please visit [lavacharts.com](http://lavacharts.com/)

---

## Installing
In your project's main ```composer.json``` file, add this line to the requirements:

  ```
  "khill/lavacharts": "3.0.x"
  ```

Run Composer to install Lavacharts:

  ```
  composer update
  ```

## Laravel Service Provider
### Laravel 5.x
Register Lavacharts in your app by adding this line to the end of the providers array in ```config/app.php```:
  ```
  'providers' => [
      ...

      Khill\Lavacharts\Laravel\LavachartsServiceProvider::class
  ],
  ```
The ```Lava::``` alias will be registered automatically via the service provider.

### Laravel 4.x
Register Lavacharts in your app by adding this line to the end of the providers array in ```app/config/app.php```:

  ```
  'providers' => array(
      ...

      "Khill\Lavacharts\Laravel\LavachartsServiceProvider"
  ),
  ```
The ```Lava::``` alias will be registered automatically via the service provider.

## Non-Laravel
If you are using Lavacharts with Composer and not in Laravel, that's fine, just make sure to:
```require 'vendor/autoload.php';``` within you project.

Create an instance of Lavacharts: ```$lava = new Khill\Lavacharts\Lavacharts;```

Replace all of the ```Lava::``` aliases in the examples, by chaining from the Lavacharts object you created.

Ex: ```$dt = $lava->DataTable();``` instead of ```$dt = Lava::DataTable();```


# Usage
The creation of charts is separated into two parts:
First, within a route or controller, you define the chart, the data table, and the customization of the output.

Second, within a view, you use one line and the library will output all the necessary javascript code for you.

## Basic Example
Here is an example of the simplest chart you can create: A line chart with one dataset and a title, no configuration.

### Controller
```
    $stocksTable = $lava->DataTable();  // Lava::DataTable() if using Laravel

    $stocksTable->addDateColumn('Day of Month')
                ->addNumberColumn('Projected')
                ->addNumberColumn('Official');

    // Random Data For Example
    for ($a = 1; $a < 30; $a++)
    {
        $rowData = [
          "2014-8-$a", rand(800,1000), rand(800,1000)
        ];

        $stocksTable->addRow($rowData);
    }
```

Arrays work for datatables as well...
```
  $stocksTable->addColumns([
    ['date', 'Day of Month'],
    ['number', 'Projected'],
    ['number', 'Official']
  ]];
```

...and for setting chart options!
```
  $lava->LineChart('Stocks', $stocksTable, ['title' => 'Stock Market Trends']);
```

## View
If you are using Laravel and the Blade templating engine, there are some nifty extensions thrown in for a cleaner view

  ```
  @linechart('Stocks', 'stocks-div');
  // Behind the scenes this just calls Lava::renderLineChart('Stocks', 'stocks-div')
  // which is an alias for the render method, seen below
  ```

Or you can use the new render method, passing in the chart type, label, and element id.

  ```
  echo Lava::render('LineChart', 'Stocks', 'stocks-div');
  ```

This is all assuming you already have a div in your page with the id "stocks-div":
```<div id="stocks-div"></div>```

If you don't have a div ready to accept the charts, add one more parameter to ```@linechart()``` or ```render()``` and it will be created for you.

Add ```true``` to for the library to create a plain div, or an array with keys ```width & height```

Example:
```
  @linechart('Stocks', 'stocks-div', true)
  // Or
  echo Lava::render('LineChart', 'Stocks', 'stocks-div', ['width'=>1024, 'height'=>768]);
```

# Changelog
The complete changelog can be found [here](https://github.com/kevinkhill/lavacharts/wiki/Changelog)


CHANGELOG.md
- 3.0.0
  - Dropping support for PHP 5.3
    - Minimum version PHP 5.4+
  - Added Dashboards
      - ChartWrappers
      - ControlWrappers
      - Added filters for Dashboard ControlWrappers
       - Category
       - ChartRange
       - DateRange
       - NumberRange
       - String
  - Chart Improvements
    - Assign DataTable and options via constructor.
    - Refactored all ConfigObject creation into the classes, no more manually instantiation.
    - Removed Event classes in favor of associative array definitions of events.
  - DataTable Improvements
    - Added support for csv file read/write and Laravel collection parsing into DataTables,
      just add the separate package to composer "khill/datatableplus":"dev-master".
      DataTablePlus extends the DataTable to add the extra functions and Lavacharts will seamlessly
      create DataTablePluses over DataTables if available via composer.
  - Added ScatterChart & TableChart
  - Added Format#formatColumn method to format datatable columns.
  - Added new formats.
   - ArrowFormat
   - BarFormat
  - lava.js has been refactored:
    - lava.get() replaced with:
      - getChart(label, callback) -> callback(Google chartObj, Lava chartObj)
        - Google chart object allows for using google's documented chart methods
        - Lava chart object contains all relevant info about the chart
         - chart, data, options, formats etc...
      - getDashboard(label, callback) -> callback(Google dashboardObj, Lava dashboardObj)
        - Google dashboard object allows for using google's documented dashboard methods
        - Lava dashboard object contains all relevant info about the dashboard
         - dashboard, control bindings, chart wrapper, data, options, etc...
    - lava.ready() used to wrap other lava.js interaction.
      - Called after all of the google jsapi is loaded and the charts are rendered.

- 2.5.7
 - AJAX data loading bugfixes

- 2.5.6
  - Fixes for AJAX chart loading

- 2.5.5
  - Blade extensions fix

- 2.5.4
  - Fixed namespace bug

- 2.5.3
  - Added column roles

- 2.5.2
  - AddedTimeOfDay columns

- 2.5.1
  - Lava.js bug fixes

- 2.5.0
  - Ajax chart loading

- 2.4.2
  - Added Lava#exists() method for checking if a chart exists

- 2.4.1
  - Added focusTarget to Line and Area charts

- 2.4.0
  - BarCharts added
  - Cleaned up code to PSR2
  - Changed from PSR0 -> PSR4

- 2.3.0
  - Added jsapi() method for manual script placement.

- 2.2.1
  - Timezone fixes
     Now they can be set with the constructor when creating a DataTable.

- 2.2.0
  - Gauge Chart support

- 2.1.0
  - Calendar Chart support

- 2.0.5
  - Updated Carbon
  - Laravel 5 compatibility

- 2.0.4
  - Multiple chart bug fixes

- 2.0.3
  - Fixing event bugs

- 2.0.2
  - Responsive charts

- 2.0.1
  - Multiple chart support

- 2.0.0
  - Its Here!

- 2.0.0-beta1
  - Passed 75% test coverage
  - Added new options to TextStyle
    - Bold
    - Italic

- 2.0.0-alpha4
  - Added Events
    - select
    - onmouseover
    - onmouseout

- 2.0.0-alpha3
  - Added DataTable column formatters
    - DateFormat
    - NumberFormat

- 2.0.0-alpha2
  - Added render method in favor of outputInto method
  - Added blade template extensions for seamless chart rendering
  - Moar tests!

- 2.0.0-alpha1
  - Refactored the main Lavacharts class to not be static anymore (yay!)
  - Moved the creation of the javascript into it's own class
  - Added a new class "Volcano" to store all the charts.
  - Modfied the charts to not staticly call the Lavacharts functions
  - DataTables are no longer magic, but applied via method chaining
  - Added render method in favor of outputInto method
  - Added blade template extensions as aliases to the render method
  - Tests tests tests!
  - Using phpcs to bring all the code up to PSR2 standards


sami.cfg.php
<?php

use Sami\Sami;
use Sami\RemoteRepository\GitHubRemoteRepository;
use Sami\Version\GitVersionCollection;
use Symfony\Component\Finder\Finder;

$dir = realpath(__DIR__.'/../../php/lavacharts/src');

$iterator = Finder::create()
    ->files()
    ->name('*.php')
    ->exclude('Laravel')
    //->exclude('Traits')
    ->in($dir);

$versions = GitVersionCollection::create($dir)
    ->add('2.5', '2.5.x Branch')
    ->add('3.0', '3.0.x Branch');

$options = [
    'theme'                => 'lava',
    'versions'             => $versions,
    'title'                => 'Lavacharts API',
    'build_dir'            => __DIR__.'/_site/api/%version%',
    'cache_dir'            => __DIR__.'/.cache/%version%',
    'remote_repository'    => new GitHubRemoteRepository('khill/lavacharts', dirname($dir)),
    'default_opened_level' => 2,
];

$sami = new Sami($iterator, $options);

$templates = $sami['template_dirs'];
$templates[] = __DIR__ . '/themes/';

$sami['template_dirs'] = $templates;

return $sami;
